File: app/src/main/java/com/examplesonly/android/worker/FcmTokenUpdater.java
package com.examplesonly.android.worker;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.examplesonly.android.account.UserDataProvider;
import com.examplesonly.android.network.Api;
import com.examplesonly.android.network.user.UserInterface;
import com.examplesonly.android.network.video.VideoInterface;

import java.util.HashMap;

import retrofit2.Response;
import timber.log.Timber;

public class FcmTokenUpdater extends Worker {

    private String token = null;

    private UserInterface userInterface;
    private UserDataProvider userDataProvider;

    public FcmTokenUpdater(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);

        userDataProvider = UserDataProvider.getInstance(getApplicationContext());
        userInterface = new Api(getApplicationContext()).getClient().create(UserInterface.class);
    }

    @NonNull
    @Override
    public Result doWork() {

        token = getInputData().getString("token");

        try {
            Response<HashMap<String, String>> response = userInterface
                    .updateFcmToken(userDataProvider.getRefreshToken(), token).execute();
            if (response.isSuccessful()) {
                Timber.e("FcmTokenUpdater %s", response.body().toString());
                return Result.success();
            } else {
                return Result.retry();
            }
        } catch (Exception e) {
            e.printStackTrace();
            return Result.retry();
        }
    }
}

File: app/src/main/java/com/examplesonly/android/worker/FcmTokenUpdater.kt
package com.examplesonly.android.worker

import android.content.Context
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import com.examplesonly.android.account.UserDataProvider
import com.examplesonly.android.network.Api
import com.examplesonly.android.network.user.UserInterface
import com.google.android.gms.tasks.Task
import com.google.firebase.messaging.FirebaseMessaging
import timber.log.Timber
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

class FcmTokenUpdater(context: Context, workerParams: WorkerParameters) : CoroutineWorker(context, workerParams) {

    private var token: String? = null
    private val userInterface: UserInterface = Api(applicationContext).client.create(UserInterface::class.java)
    private val userDataProvider: UserDataProvider = UserDataProvider.getInstance(applicationContext)

    override suspend fun doWork(): Result {
        token = inputData.getString("token")

        // If token is not passed, Get it from FirebaseMessaging
        if (token == null) {
            val task = FirebaseMessaging.getInstance().token.execute()

            if (!task.isSuccessful)
                return Result.retry()

            token = task.result
        }

        return token?.let { publishToken(it) } ?: run {
            Result.retry()
        }
    }

    private fun publishToken(token: String): Result {
        return try {
            val response = userInterface
                    .updateFcmToken(userDataProvider.refreshToken, token).execute()
            if (response.isSuccessful) {
                Timber.e("FcmTokenUpdater %s", response.body().toString())
                Result.success()
            } else {
                Result.retry()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            Result.retry()
        }
    }

    private suspend fun <T> Task<T>.execute() = suspendCoroutine<Task<T>> { cont ->
        addOnCompleteListener {
            cont.resume(it)
        }
    }
}